### PR TITLE
add quantile regression

### PR DESCRIPTION
**Pull request by shihchengli**
## Description
The implementation of conformal quantile prediction for v2 outputs the upper and lower bounds of a prediction from the neural network. I have transformed this into a mean and interval format to match the other methods, where the first output is the prediction results. We can also revert to the original design, but we would then need to additionally take care of the metrics calculation for this case. I am open to any suggestions.

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/nn/predictors.py [hunk 1]
```diff
@@ -203,6 +204,30 @@ def forward(self, Z: Tensor) -> Tensor:
...
+        upper_pred = self.output_transform(upper_pred)
+
+        mean = (lower_pred + upper_pred) / 2
+        interval = upper_pred - lower_pred
+
+        return torch.cat((mean, interval), 1)
+
+    def train_step(self, Z: Tensor) -> Tensor:
+        Y = super().forward(Z)
+        mean, interval = torch.chunk(Y, self.n_targets, 1)
```
Comment (davidegraff) [inline chemprop/nn/predictors.py:226]: isn't the output from `super().forward()` a tensor containing the lower and upper bound of prediction? At least that's what L214 implies.

File: chemprop/nn/loss.py [hunk 2]
```diff
@@ -348,3 +349,30 @@ def _calc_unreduced_loss(self, preds: Tensor, targets: Tensor, mask: Tensor, *ar
...
+class QuantileLoss(LossFunction):
+    def __init__(self, task_weights: ArrayLike = 1.0, alpha: float = 0.1):
+        super().__init__(task_weights)
+
+        self.alpha = alpha
+        self.tau_lower = alpha / 2
+        self.tau_upper = 1 - alpha / 2
+
+    def _calc_unreduced_loss(self, preds: Tensor, targets: Tensor, mask: Tensor, *args) -> Tensor:
+        mean, interval = torch.chunk(preds, 2, 1)
```
Comment (davidegraff) [inline chemprop/nn/loss.py:364]: broader question here: because we don't currently support uncertainty prediction, we can change the current object model. Currently, we output a tensor of shape $(b, t*s)$, where $b$ is the batch size, $t$, the number of tasks, and $s$ the number of uncertainty targets. It makes it a little awkward to unpack predictions, as is the case here, where we have to chunk the raw tensor. What do we think about changing the output shape of uncertainty-based model to $(b, t, s)$, so that we have a separate dimension of the tensor for each concept: input, task, and uncertainty parameter?

File: chemprop/nn/predictors.py (hunk 1 above)
Comment (shihchengli) [reply chemprop/nn/predictors.py:226]: Yes, it is a tensor containing both lower and upper bounds.

File: chemprop/nn/loss.py (hunk 2 above)
Comment (shihchengli) [reply chemprop/nn/loss.py:364]: The number of outputs for the FFN is $t * s$. Thus, I think we have to chunk the tensor in forward() of the predictor. I guess you mean to replace the output of `torch.cat((mean, v, alpha, beta), 1)` with `torch.stack((mean, v, alpha, beta), dim=2)` in the case of `EvidentialFFN`? Anyhow, I think it would be a good idea to separate those prediction results into separate dimensions. It can help us more easily understand the output of MveFFN, EvidentialFFN, BinaryDirichletFFN, and QuantileFFN.

Comment (davidegraff) [reply chemprop/nn/loss.py:364]: Exactly, placing separate concepts into separate dimensions, ex: $\hat y_{ijk}$, where $i$ indexes the input, $j$, the input, and $k$ the uncertainty parameter. So if someone wants only the mean predictions, I think it's clearer to write `preds[:, :, 0]` rather than `preds[:, ::s]`, where `s` is the number of uncertainty parameters and needs to be known somehow. It's not a performance difference, but I just think it's clearer.

File: chemprop/nn/predictors.py (hunk 1 above)
Comment (davidegraff) [reply chemprop/nn/predictors.py:226]: so this should be
```suggestion
        lower, upper = torch.chunk(Y, self.n_targets, 1)
```
then, right?

Comment (shihchengli) [reply chemprop/nn/predictors.py:226]: Oops! I intended to use `Y = self.forward(Z)` instead of `Y = super().forward(Z)` here. Thus, the output is mean and interval, which are further used to reconstruct the upper and lower bounds in `QuantileLoss` to calculate the loss.

File: chemprop/nn/loss.py [hunk 3]
```diff
@@ -348,3 +349,30 @@ def _calc_unreduced_loss(self, preds: Tensor, targets: Tensor, mask: Tensor, *ar
...
+        lower_pred = mean + interval / 2
+        upper_pred = mean - interval / 2
+
+        error_lower = targets - lower_pred
+        error_upper = targets - upper_pred
+
+        loss_lower = torch.max(self.tau_lower * error_lower, (self.tau_lower - 1) * error_lower)
+        loss_upper = torch.max(self.tau_upper * error_upper, (self.tau_upper - 1) * error_upper)
+
+        return loss_lower + loss_upper
```
[the remark this answers is not in the document]
Comment (davidegraff) [reply chemprop/nn/loss.py:375]: Also, double checking: are L366-7 correct?

File: chemprop/nn/loss.py [hunk 4]
```diff
@@ -348,3 +349,30 @@ def _calc_unreduced_loss(self, preds: Tensor, targets: Tensor, mask: Tensor, *ar
...
+
+
+@LossFunctionRegistry.register(["quantile", "pinball"])
+class QuantileLoss(LossFunction):
+    def __init__(self, task_weights: ArrayLike = 1.0, alpha: float = 0.1):
+        super().__init__(task_weights)
+
+        self.alpha = alpha
+        self.tau_lower = alpha / 2
+        self.tau_upper = 1 - alpha / 2
```
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/nn/loss.py:361]: I think it will be cast into a tensor via `task_weights = torch.as_tensor(task_weights, dtype=torch.float).view(1, -1)` at L39.

File: chemprop/nn/loss.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/nn/loss.py:375]: @davidegraff thanks for refactoring this and for the detailed explanation of this part. I will incorporate it. Also, thanks for catching the bug at L366-7.

File: chemprop/nn/loss.py (hunk 2 above)
Comment (hwpang) [reply chemprop/nn/loss.py]: I love this idea! It makes handling predictions a lot easierm and also make the code easier to understand. I will work on https://github.com/chemprop/chemprop/pull/959 accordingly.

**Comment by shihchengli** [comment]
This PR depends on #959

File: chemprop/nn/predictors.py [hunk 5]
```diff
@@ -203,6 +204,30 @@ def forward(self, Z: Tensor) -> Tensor:
...
+        mean = (lower_bound + upper_bound) / 2
+        interval = upper_bound - lower_bound
+
+        return torch.cat((mean, interval), 1)
+
+    def train_step(self, Z: Tensor) -> Tensor:
+        Y = self.forward(Z)
+        mean, interval = torch.chunk(Y, self.n_targets, 1)
+
+        return torch.cat((mean, interval), 1)
```
Comment (KnathanM) [inline chemprop/nn/predictors.py:228]: I'm confused why you chunk the mean and interval and then immediately cat them again.

File: chemprop/nn/loss.py [hunk 6]
```diff
@@ -348,3 +349,31 @@ def _calc_unreduced_loss(self, preds: Tensor, targets: Tensor, mask: Tensor, *ar
...
+        super().__init__(task_weights)
+        self.alpha = alpha
+
+        bounds = torch.tensor([-1 / 2, 1 / 2]).view(-1, 1, 1)
+        tau = torch.tensor([[alpha / 2, 1 - alpha / 2], [alpha / 2 - 1, -alpha / 2]]).view(
+            2, 2, 1, 1
+        )
+
+        self.register_buffer("bounds", bounds)
+        self.register_buffer("tau", tau)
```
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/nn/loss.py]: Thanks good!

File: chemprop/nn/metrics.py [hunk 7]
```diff
@@ -207,3 +209,8 @@ class SIDMetric(SIDLoss, Metric):
 @MetricRegistry.register("wasserstein")
 class WassersteinMetric(WassersteinLoss, Metric):
     pass
+
+
+@MetricRegistry.register("quantile")
+class QuantileMetric(QuantileLoss, Metric):
+    pass
```
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/nn/metrics.py:216]: MSE can be used as the metric. However, it seems a bit odd to me not to use quantile loss as the val_loss. In general, I think we should support using the criterion to calculate `val_loss`. Currently, we can bypass this issue by not specifying metrics when creating the MPNN object so it still use the criterion to calcaulte `val_loss`.

[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/nn/metrics.py:216]: I can remove this from the metrics in this PR. As for `EvidentialLoss` and `MveFFN`, we currently set the default to `MSEMetric`.

File: chemprop/nn/predictors.py (hunk 5 above)
Comment (shihchengli) [reply chemprop/nn/predictors.py:228]: Good catch! I will fix it.

**Comment by shihchengli** [comment]
This PR is ready for another round of review. I change the name of `ConformalRegressionPredictor` and `ConformalQuantileRegressionPredictor` into `QuantileRegressionPredictor`, as the interval is derived from quantile regression without any calibration. Additionally, I have added the code for `QuantileRegressionPredictor`.

File: chemprop/nn/loss.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/nn/loss.py]: After removing them from the buffer, we have to manually move them to the correct device. Therefore, I create these two objects each time the loss calculation is called. Does that sound okay?

File: chemprop/nn/loss.py [hunk 8]
```diff
@@ -348,3 +349,30 @@ def _calc_unreduced_loss(self, preds: Tensor, targets: Tensor, mask: Tensor, *ar
...
+
+    def _calc_unreduced_loss(self, preds: Tensor, targets: Tensor, mask: Tensor, *args) -> Tensor:
+        mean, interval = torch.unbind(preds, dim=-1)
+
+        device = interval.device
+        self.bounds = torch.tensor([-1 / 2, 1 / 2], device=device).view(2, 1, 1)
+        self.tau = torch.tensor(
+            [[self.alpha / 2, 1 - self.alpha / 2], [self.alpha / 2 - 1, -self.alpha / 2]],
+            device=device,
+        ).view(2, 2, 1, 1)
```
Comment (davidegraff) [inline chemprop/nn/loss.py:368]: if `alpha` is only used insofar as it defines `self.tau`, why are we recalculating it in _every_ call to `_calc_unreduced_loss()`? This function is called **every batch**, so we're repeating the same work $b$ times every epoch. I don't think it matters inasmuch as this work is minimal, but I think it actually harms readability IMO.

File: chemprop/nn/predictors.py [hunk 9]
```diff
@@ -203,6 +204,27 @@ def forward(self, Z: Tensor) -> Tensor:
...
 
 
+@PredictorRegistry.register("regression-quantile")
+class QuantileFFN(RegressionFFN):
+    n_targets = 2
+    _T_default_criterion = QuantileLoss
+
+    def forward(self, Z: Tensor) -> Tensor:
+        Y = super().forward(Z)
+        lower_bound, upper_bound = torch.chunk(Y, self.n_targets, 1)
```
Comment (davidegraff) [inline chemprop/nn/predictors.py:214]: I though uncertainty targets were in the final dimension now?

File: chemprop/nn/loss.py (hunk 8 above)
Comment (shihchengli) [reply chemprop/nn/loss.py:368]: Originally, I registered `bound` and `tau` as buffers so they would be saved in the state dict. However, since they can be recreated from `alpha`, I removed their registration as buffers. Do you have any suggestions for this? Maybe add an if statement in `_calc_unreduced_loss` to check if the devices are consistent. If not, move them to the correct device?

File: chemprop/nn/predictors.py (hunk 9 above)
Comment (shihchengli) [reply chemprop/nn/predictors.py:214]: The quantile regression will predict the $\alpha$ and $1 - \alpha$ quantiles for a given input. These two quantiles are then used to calculate the mean as the prediction and the interval as the uncertainty. The forward function in `QuantileFFN` will still return the uncertainty targets in the final dimension (see L222).

File: chemprop/nn/loss.py (hunk 8 above)
Comment (davidegraff) [reply chemprop/nn/loss.py:368]: What's wrong with registering them as buffers? I don't see a problem with storing these as buffers and then _also_ storing `alpha` even though it's not actually used and thus "wasting" memory (but it's really only 8 bytes of extra storage _per object_, and we'll likely only have 1 object alive at any time, so the impact is negligible.

File: chemprop/nn/predictors.py (hunk 9 above)
Comment (davidegraff) [reply chemprop/nn/predictors.py:214]: oh yes my bad- thanks for the clarification.

File: chemprop/nn/loss.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/nn/loss.py]: Since removing them as buffers would require additional code to handle device movement (https://github.com/chemprop/chemprop/pull/963#discussion_r1787943094), perhaps we should stick with the original design?

File: chemprop/uncertainty/predictor.py [hunk 10]
```diff
@@ -123,17 +124,16 @@ def __call__(
...
-    ) -> tuple[Tensor, Tensor]:
-        return
+        individual_preds = []
+        for model in models:
+            predss = trainer.predict(model, dataloader)
+            individual_preds.append(torch.concat(predss, 0))
+        stacked_preds = torch.stack(individual_preds).float()
+        mean = stacked_preds[..., 0]
+        interval = torch.mean(stacked_preds[..., 1], dim=0)
+        return mean, interval
```
Comment (shihchengli) [inline chemprop/uncertainty/predictor.py:139]: Should this method return variance as the uncertainty to match the other methods (i.e., ensemble, dropout, MVE, evidential)?

File: chemprop/nn/metrics.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (KnathanM) [reply chemprop/nn/metrics.py]: In my metric overhaul PR I change how val_loss is calculated to reuse the loss function, so we won't need a QuantileMetric. However you deal with this for now should be fine.

File: chemprop/nn/predictors.py [hunk 11]
```diff
@@ -203,6 +204,27 @@ def forward(self, Z: Tensor) -> Tensor:
...
+        lower_bound = self.output_transform(lower_bound)
+        upper_bound = self.output_transform(upper_bound)
+
+        mean = (lower_bound + upper_bound) / 2
+        interval = upper_bound - lower_bound
+
+        return torch.stack((mean, interval), dim=2)
+
+    def train_step(self, Z: Tensor) -> Tensor:
+        return self.forward(Z)
```
Comment (KnathanM) [inline chemprop/nn/predictors.py:225]:
```suggestion
    train_step = foward
```

File: chemprop/uncertainty/predictor.py (hunk 10 above)
Comment (KnathanM) [reply chemprop/uncertainty/predictor.py:139]: I'm not an expert, but it seems like we would have to assume a distribution to go from interval to variance, and I am not sure we can do that.

File: tests/cli/test_cli_regression_mol+mol.py [hunk 12]
```diff
@@ -62,7 +62,7 @@ def test_train_quick(monkeypatch, data_path):
         "--show-individual-scores",
     ]
 
-    task_types = ["", "regression-mve", "regression-evidential"]
```
Comment (KnathanM) [inline tests/cli/test_cli_regression_mol+mol.py:65]: If the task type is regression-evidential, we set the evidential-regularization to 0.2. If task type is regression-quantile, should we set alpha to a value?

File: chemprop/uncertainty/predictor.py (hunk 10 above)
Comment (davidegraff) [reply chemprop/uncertainty/predictor.py]: I mean we train these things using max likelihood estimation from a unit normal gaussian, so it seems reasonable to assume a gaussian for this purpose

File: tests/cli/test_cli_regression_mol+mol.py (hunk 12 above)
Comment (shihchengli) [reply tests/cli/test_cli_regression_mol+mol.py:65]: The default value for `alpha` is 0.1, so I think it is fine not testing the other values.

File: chemprop/uncertainty/predictor.py (hunk 10 above)
Comment (shihchengli) [reply chemprop/uncertainty/predictor.py]: The quantile loss function minimizes a weighted sum of absolute residuals, so I think it is not a maximum likelihood method. I will revert the change to still return the intervals as uncertainty.

**Review by KnathanM — APPROVED** [review]
